File: py/aoc2023/day9.py
"""
Day 9: Mirage Maintenance
"""

SAMPLE_INPUT = """
0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
"""


def _extrapolate(nums):
    if any(nums):
        return nums[-1] + _extrapolate([y - x for x, y in zip(nums, nums[1:])])
    return 0


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    114
    """
    return sum(
        _extrapolate([int(word) for word in line.split() if word])
        for line in data.splitlines()
    )


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    2
    """
    return sum(
        _extrapolate([int(word) for word in line.split() if word][::-1])
        for line in data.splitlines()
    )


parts = (part1, part2)

File: rs/src/day9.rs
fn predict(nums: &[i32]) -> i32 {
    if nums.iter().any(|&x| x != 0) {
        nums.last().unwrap()
            + predict(
                &nums
                    .iter()
                    .zip(nums.iter().skip(1))
                    .map(|(x, y)| y - x)
                    .collect::<Vec<_>>(),
            )
    } else {
        0
    }
}

pub fn part1(data: &str) -> i32 {
    data.lines()
        .map(|line| {
            predict(
                &line
                    .split_whitespace()
                    .filter_map(|s| s.parse::<i32>().ok())
                    .collect::<Vec<_>>(),
            )
        })
        .sum()
}

pub fn part2(data: &str) -> i32 {
    data.lines()
        .map(|line| {
            predict(
                &line
                    .split_whitespace()
                    .rev()
                    .filter_map(|s| s.parse::<i32>().ok())
                    .collect::<Vec<_>>(),
            )
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        0 3 6 9 12 15
        1 3 6 10 15 21
        10 13 16 21 30 45
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(114, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(2, part2(EXAMPLE));
    }
}
